pkg/etcdenvvar/envvarcontroller.go

```diff
@@ -128,15 +130,58 @@ func (c *EnvVarController) checkEnvVars() error {
 		return err
 	}
 
-	currEnvVarMap, err := getEtcdEnvVars(envVarContext{
+	ctx := envVarContext{
 		targetImagePullSpec:  c.targetImagePullSpec,
 		spec:                 *operatorSpec,
 		status:               *operatorStatus,
 		configmapLister:      c.configmapLister,
 		nodeLister:           c.nodeLister,
 		infrastructureLister: c.infrastructureLister,
 		networkLister:        c.networkLister,
-	})
+	}
+
+	bootstrapComplete, err := isBootstrapComplete(c.configmapLister, c.operatorClient)
+	if err != nil {
+		return fmt.Errorf("couldn't determine bootstrap status: %w", err)
+	}
+	isUnsupportedUnsafeEtcd, err := ceohelpers.IsUnsupportedUnsafeEtcd(&ctx.spec)
+	if err != nil {
+		return fmt.Errorf("couldn't determine etcd unsupported override status: %w", err)
+	}
+	// TODO: determine this through a state check
+	isManagedByAssistedInstaller := true
+	nodeCount := len(ctx.status.NodeStatuses)
+
+	// There are three discrete validation paths to enforce HA invariants
+	// depending on whether the unsupported/unsafe config is set, whether the
+	// cluster is managed by assisted installer, and the default configuration.
+	switch {
+	case isUnsupportedUnsafeEtcd:
+		// TODO: this is to help make sure assisted installer can run in a supported config.
+		// TODO: can remove this after integration testing.
+		if isManagedByAssistedInstaller {
+			return fmt.Errorf("assisted installer is running with an unsupported etcd configuration")
+		}
+		// When running in an unsupported/unsafe configuration, no guarantees are
+		// made once a single node is available.
+		if nodeCount < 1 {
+			return fmt.Errorf("at least one node is required to have a valid configuration")
+		}
+	case isManagedByAssistedInstaller:
+		// When managed by assisted installer, tolerate unsafe conditions only up
+		// until bootstrap is complete, and then enforce as in the supported case.
+		if nodeCount < 3 && bootstrapComplete {
```

Comment:
In the assisted installer flow there is a time gap between bootstrap complete and the time the 3rd master joins.
(the bootstrap node pivots to be the 3rd master once bootstrap is completed)

Reply:
bootstrap being bootkube service running on the bootstrap node

Reply:
I agree that this decision has implications. technically speaking the install-config would represent 2 master replicas. In which case install complete should require 2 master nodes. Then scale up to 3 would be a secondary process to the install. Understanding a point in time where we are installComplete is important.

> After install status is `complete` the operator will go into a degraded state until the cluster has 3 or more master nodes.

So this allows a cluster to achieve install complete of less than 3 master nodes but not tolerate less than three after that point. So ideally you would resolve the scaling before install-complete but it would not be required. In the case where no 3rd master ever joined the cluster, it would remain degraded with clear message.

[1] https://github.com/openshift/enhancements/pull/480